FWCore/Framework/src/EventProcessor.cc

```diff
@@ -1294,159 +1294,149 @@ namespace edm {
           // lumi is done and no longer needs its EventSetup IOVs.
-          espController_->eventSetupForInstanceAsync(
-              iSync, queueLumiWorkTaskHolder, status->endIOVWaitingTasks(), status->eventSetupImpls());
+          espController->eventSetupForInstanceAsync(
+              iSync, task, status->endIOVWaitingTasks(), status->eventSetupImpls());
           sentry.completedSuccessfully();
         } catch (...) {
-          queueLumiWorkTaskHolder.doneWaiting(std::current_exception());
+          task.doneWaiting(std::current_exception());
         }
-        queueWhichWaitsForIOVsToFinish_.pause();
-      });
-
-    } else {
-      queueWhichWaitsForIOVsToFinish_.pause();
-
-      // This holder will be used to wait until the EventSetup IOVs are ready
-      WaitingTaskHolder queueLumiWorkTaskHolder{*iHolder.group(), queueLumiWorkTask};
-      // Caught exception is propagated via WaitingTaskHolder
-      CMS_SA_ALLOW try {
-        SendSourceTerminationSignalIfException sentry(actReg_.get());
+      };
+      if (espController_->doWeNeedToWaitForIOVsToFinish(iSync)) {
+        // We only get here inside this block if there is an EventSetup
+        // module not able to handle concurrent IOVs (usually an ESSource)
+        // and the new sync value is outside the current IOV of that module.
+        auto group = nextTask.group();
+        queueWhichWaitsForIOVsToFinish_.push(
+            *group, [this, task = std::move(nextTask), iSync, status, asyncEventSetup]() mutable {
+              asyncEventSetup(
+                  actReg_.get(), espController_.get(), queueWhichWaitsForIOVsToFinish_, std::move(task), status, iSync);
+            });
+      } else {
+        asyncEventSetup(
+            actReg_.get(), espController_.get(), queueWhichWaitsForIOVsToFinish_, std::move(nextTask), status, iSync);
+      }
+    }) | chain::then([this, status](std::exception_ptr const* iPtr, auto nextTask) {
+      //the call to doneWaiting will cause the count to decrement
+      auto copyTask = nextTask;
+      if (iPtr) {
+        nextTask.doneWaiting(*iPtr);
+      }
+      auto group = copyTask.group();
+      lumiQueue_->pushAndPause(
+          *group, [this, task = std::move(copyTask), status](edm::LimitedTaskQueue::Resumer iResumer) mutable {
+            if (task.taskHasFailed()) {
+              status->resetResources();
+              return;
+            }
 
-        // Pass in iSync to let the EventSetup system know which run and lumi
-        // need to be processed and prepare IOVs for it.
-        // Pass in the endIOVWaitingTasks so the lumi can notify them when the
-        // lumi is done and no longer needs its EventSetup IOVs.
-        espController_->eventSetupForInstanceAsync(
-            iSync, queueLumiWorkTaskHolder, status->endIOVWaitingTasks(), status->eventSetupImpls());
-        sentry.completedSuccessfully();
+            status->setResumer(std::move(iResumer));
+
+            auto group = task.group();
```

Comment:
NOTE: I think in a latter PR this routine should be subdivided into smaller functions which would be called by this routine.